src/Microsoft.AspNetCore.Routing/Template/TemplateBinder.cs

```diff
@@ -388,47 +394,71 @@ private static bool IsRoutePartNonEmpty(object routePart)
         private struct TemplateBindingContext
         {
             private readonly RouteValueDictionary _defaults;
-            private readonly RouteValueDictionary _acceptedValues;
+            private RouteValueDictionary _acceptedValues;
+            private readonly RouteValueDictionary _values;
 
-            public TemplateBindingContext(RouteValueDictionary defaults)
+            public TemplateBindingContext(RouteValueDictionary defaults, RouteValueDictionary values)
             {
                 _defaults = defaults;
+                _values = values;
 
-                _acceptedValues = new RouteValueDictionary();
+                // Perf: In most of the common cases, AcceptedValues is same as the values.
+                // So, we lazy initialize a new RouteValueDictionary only when they are different.
+                _acceptedValues = null;
             }
 
             public RouteValueDictionary AcceptedValues
             {
-                get { return _acceptedValues; }
+                get { return _acceptedValues ?? _values; }
             }
 
-            public void Accept(string key, object value)
+            public void AcceptIfNonEmpty(string key, object value)
             {
-                if (!_acceptedValues.ContainsKey(key))
+                if (IsRoutePartNonEmpty(value))
                 {
-                    _acceptedValues.Add(key, value);
+                    if (NeedsValue(key))
+                    {
+                        if (_acceptedValues == null)
+                        {
+                            // Perf: Lazy initialize acceptedValues only if a new value needs to be added.
+                            _acceptedValues = new RouteValueDictionary();
+                        }
+
+                        _acceptedValues.Add(key, value);
+                    }
+                }
+                else if (_acceptedValues == null)
+                {
+                    // 'values' dictionary contains an empty value which should not be accepted.
+                    // Eg: The template 'foo/{controller}' with 'values' { "controller", "" }
+                    // should generate 'null'.
+                    _acceptedValues = new RouteValueDictionary();
```

Comment:
I think we really need to document that this is stateful.

Also, I think ordering is an issue here - like if I "see" `{ "controller", "" }` first this seems correct. If I see `{ "action", "Index" }` and then `{ "controller", "" }` it seems like this would be incorrect because `_acceptedValues` will be missing the value for `action`.

The more I think about this change the less I like it.